velox/functions/prestosql/GeometryUtils.h

```diff
@@ -0,0 +1,685 @@
+
+  static void skipEsriType(SliceInput& input) {
+    input.skip(4); // Esri type is an integer
+  }
+
+  static void skipEnvelope(SliceInput& input) {
+    input.skip(4 * 8); // Envelopes are 4 doubles (minX, minY, maxX, maxY)
+  }
+
+  static geos::geom::Coordinate readCoordinate(SliceInput& input) {
+    double x = input.readDouble();
+    double y = input.readDouble();
+    return {x, y};
+  }
+
+  static std::unique_ptr<geos::geom::CoordinateSequence> readCoordinates(
+      SliceInput& input,
+      int count) {
+    auto coords = std::make_unique<geos::geom::CoordinateSequence>(count, 2);
+    for (int i = 0; i < count; ++i) {
+      coords->setAt(readCoordinate(input), i);
+    }
+    return coords;
+  }
+
+  static std::unique_ptr<geos::geom::Point> readPoint(SliceInput& input) {
+    geos::geom::Coordinate coordinate = readCoordinate(input);
+    if (std::isnan(coordinate.x) || std::isnan(coordinate.y)) {
+      return GEOMETRY_FACTORY->createPoint();
+    }
+    return std::unique_ptr<geos::geom::Point>(
+        GEOMETRY_FACTORY->createPoint(coordinate));
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> readMultiPoint(
+      SliceInput& input) {
+    skipEsriType(input);
+    skipEnvelope(input);
+    int pointCount = input.readInt();
+    auto coords = readCoordinates(input, pointCount);
+    std::vector<std::unique_ptr<geos::geom::Point>> points;
+    for (size_t i = 0; i < coords->size(); ++i) {
+      points.push_back(
+          std::unique_ptr<geos::geom::Point>(GEOMETRY_FACTORY->createPoint(
+              geos::geom::Coordinate(coords->getX(i), coords->getY(i)))));
+    }
+    return GEOMETRY_FACTORY->createMultiPoint(std::move(points));
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> readPolyline(
+      SliceInput& input,
+      bool multiType) {
+    skipEsriType(input);
+    skipEnvelope(input);
+    int partCount = input.readInt();
+
+    if (partCount == 0) {
+      if (multiType) {
+        return GEOMETRY_FACTORY->createMultiLineString();
+      }
+      return GEOMETRY_FACTORY->createLineString();
+    }
+
+    int pointCount = input.readInt();
+    std::vector<int> startIndexes(partCount);
+
+    for (int i = 0; i < partCount; ++i) {
+      startIndexes[i] = input.readInt();
+    }
+
+    std::vector<int> partLengths(partCount);
+    if (partCount > 1) {
+      partLengths[0] = startIndexes[1];
+      for (int i = 1; i < partCount - 1; ++i) {
+        partLengths[i] = startIndexes[i + 1] - startIndexes[i];
+      }
+    }
+    partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
+
+    std::vector<std::unique_ptr<geos::geom::LineString>> lineStrings;
+    for (int i = 0; i < partCount; ++i) {
+      lineStrings.push_back(GEOMETRY_FACTORY->createLineString(
+          readCoordinates(input, partLengths[i])));
+    }
+
+    if (multiType) {
+      return GEOMETRY_FACTORY->createMultiLineString(std::move(lineStrings));
+    }
+
+    if (lineStrings.size() != 1) {
+      VELOX_USER_FAIL(
```

Comment:
please fix: This would be a system error, not a user error.  This meant that _we_ messed up this serialization.